File: C#/P2.2/Generator/MathFunctions.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Navigation;

namespace P2._2.Generator
{
    static class MF
    {
        static Random r = new Random(1337);
        public static (int, int) UnitPointToMapCoords(float x, float y, int mapwidth, int mapheight) {
            return ((int)(x * (mapwidth - 1)), (int)(y * (mapheight - 1)));
        }

        public static (float, float) MapCoordsToUnitPoint(float x, float y, int mapwidth, int mapheight) {
            return ((mapwidth - 1) / x, (mapheight - 1) / y);
        }

        public static (float, float) RandomPointInSquare() {
            return ((float)r.NextDouble(), (float)r.NextDouble());
        }

        public static Random Random {
            get => r;
        }
    }

    public class LineSegment
    {
        public (float x, float y) p0 { get; }
        public (float x, float y) p1 { get; }
        public float length { get; }
        public float dir { get; }
        public int id { get; }

        public LineSegment((float, float) p0, (float, float) p1, int id) {
            this.p0 = p0;
            this.p1 = p1;
            (float, float) rls = (this.p1.x - this.p0.x, this.p1.y - this.p0.y);
            length = (float)Math.Sqrt(rls.Item1 * rls.Item1 + rls.Item2 * rls.Item2);
            dir = (float)Math.Atan2(rls.Item2, rls.Item1);
            this.id = id;
        }

        public override string ToString() {
            return $"({p0.x},{p0.y}), ({p1.x},{p1.y}), l:{length}, d:{(180 / Math.PI) * dir}";
        }
    }
}

File: Python/math_functions.py
import random
import math

class LineSegment:
    def __init__(self, p0, p1, id):
        self._p0 = p0
        self._p1 = p1
        rls = (p1[0] - p0[0], p1[1] - p0[1]) # relative line segment from END to START endpoint
        self._length = math.sqrt(rls[0] * rls[0] + rls[1] * rls[1])
        self._dir = math.atan2(rls[1], rls[0])
        self._id = id
    
    def __str__(self):
        return str(self._p0) + ", " + str(self._p1) + ", l = " + str(self._length) + ", d = " + str(math.degrees(self._dir))

# maps a point on a unit square to cell coordinates on a grid map
def UnitPointToMapCoords(x, y, mapwidth, mapheight):
    return (int(x * (mapwidth - 1)), int(y * (mapheight - 1)))

# maps cell coordinates on a grid map to a point on a unit square
def MapCoordsToUnitPoint(x, y, mapwidth, mapheight):
    return ((mapwidth - 1) / x, (mapheight - 1) / y)

def RandomPointInSquare():
    x = random.random()
    y = random.random()
    return (x, y)